# Student Do: Three-Year Loans

The higher-ups at Harold's firm want to know how to better target customers who are seeking three-year loans. Harold's manager has asked him to review a compilation of loan data, and then filter out the necessary data to generate insights about customers who have been granted three-year loans.

Follow the instructions to help Harold answer his manager's questions.

## Instructions

Complete the following steps.

### Import the necessary libraries and dependencies.

In [1]:
# Import libraries and dependencies
import pandas as pd
from pathlib import Path

### Read CSV in as DataFrame

In [2]:
# Set the path
file_path = Path("../Resources/loans.csv")

# Read in the CSV as a DataFrame
loans_csv = pd.read_csv(file_path)

### Show the First 10 Records

In [3]:
# Retrieve rows with index 0 up to 10 (not including)
loans_csv.iloc[0:10]

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  installment  \
0      10000        10000          10000.0  36 months    0.1033       324.23   
1       4000         4000           4000.0  36 months    0.2340       155.68   
2       5000         5000           5000.0  36 months    0.1797       180.69   
3       9600         9600           9600.0  36 months    0.1298       323.37   
4       2500         2500           2500.0  36 months    0.1356        84.92   
5      30000        30000          30000.0  60 months    0.1894       777.23   
6      23000        23000          23000.0  60 months    0.2089       620.81   
7      32075        32075          32075.0  60 months    0.1180       710.26   
8       8000         8000           8000.0  36 months    0.2340       311.35   
9      10000        10000          10000.0  60 months    0.1992       264.50   

  grade sub_grade           emp_title emp_length home_ownership  annual_inc  \
0     B        B1                 NaN   < 1 year       MORTGAGE    280000.0   
1     E        E1            Security    3 years           RENT     90000.0   
2     D        D1      Administrative    6 years       MORTGAGE     59280.0   
3     B        B5                 NaN        NaN       MORTGAGE     35704.0   
4     C        C1                Chef  10+ years           RENT     55000.0   
5     D        D2         Postmaster   10+ years       MORTGAGE     90000.0   
6     D        D4            Operator    5 years           RENT     68107.0   
7     B        B4  Nursing Supervisor  10+ years       MORTGAGE    150000.0   
8     E        E1             Manager  10+ years            OWN     43000.0   
9     D        D3    Material Handler  10+ years       MORTGAGE     80000.0   

  verification_status issue_d loan_status pymnt_plan             purpose  
0        Not Verified  Dec-18     Current          n  debt_consolidation  
1     Source Verified  Dec-18     Current          n  debt_consolidation  
2     Source Verified  Dec-18     Current          n  debt_consolidation  
3        Not Verified  Dec-18     Current          n    home_improvement  
4        Not Verified  Dec-18     Current          n  debt_consolidation  
5     Source Verified  Dec-18     Current          n  debt_consolidation  
6     Source Verified  Dec-18     Current          n  debt_consolidation  
7        Not Verified  Dec-18     Current          n         credit_card  
8     Source Verified  Dec-18     Current          n  debt_consolidation  
9        Not Verified  Dec-18     Current          n  debt_consolidation

### View Summary Statistics

In [4]:
# Describe summary statistics for csv data
loans_csv.describe(include="all")

loan_amnt    funded_amnt  funded_amnt_inv       term  \
count   128412.000000  128412.000000    128412.000000     128412   
unique            NaN            NaN              NaN          2   
top               NaN            NaN              NaN  36 months   
freq              NaN            NaN              NaN      88179   
mean     15971.321021   15971.321021     15968.498166        NaN   
std      10150.384233   10150.384233     10152.168970        NaN   
min       1000.000000    1000.000000       725.000000        NaN   
25%       8000.000000    8000.000000      8000.000000        NaN   
50%      14000.000000   14000.000000     14000.000000        NaN   
75%      21600.000000   21600.000000     21600.000000        NaN   
max      40000.000000   40000.000000     40000.000000        NaN   

             int_rate    installment   grade sub_grade emp_title emp_length  \
count   128412.000000  128412.000000  128412    128412    107465     116708   
unique            NaN            NaN       7        35     43892         11   
top               NaN            NaN       A        A4   Teacher  10+ years   
freq              NaN            NaN   38011      9770      2090      38826   
mean         0.129302     463.259353     NaN       NaN       NaN        NaN   
std          0.050904     285.718256     NaN       NaN       NaN        NaN   
min          0.060000      30.480000     NaN       NaN       NaN        NaN   
25%          0.084600     253.630000     NaN       NaN       NaN        NaN   
50%          0.118000     382.905000     NaN       NaN       NaN        NaN   
75%          0.161400     622.680000     NaN       NaN       NaN        NaN   
max          0.309900    1618.240000     NaN       NaN       NaN        NaN   

       home_ownership    annual_inc verification_status issue_d loan_status  \
count          128412  1.284120e+05              128412  128412      128412   
unique              4           NaN                   3       3           7   
top          MORTGAGE           NaN        Not Verified  Oct-18     Current   
freq            63490           NaN               58350   46305      119540   
mean              NaN  8.279733e+04                 NaN     NaN         NaN   
std               NaN  1.082985e+05                 NaN     NaN         NaN   
min               NaN  0.000000e+00                 NaN     NaN         NaN   
25%               NaN  4.705800e+04                 NaN     NaN         NaN   
50%               NaN  6.800000e+04                 NaN     NaN         NaN   
75%               NaN  9.900000e+04                 NaN     NaN         NaN   
max               NaN  9.757200e+06                 NaN     NaN         NaN   

       pymnt_plan             purpose  
count      128412              128412  
unique          1                  12  
top             n  debt_consolidation  
freq       128412               70603  
mean          NaN                 NaN  
std           NaN                 NaN  
min           NaN                 NaN  
25%           NaN                 NaN  
50%           NaN                 NaN  
75%           NaN                 NaN  
max           NaN                 NaN

### Filter Columns and Create Subset DataFrame

In [5]:
# Filter the DataFrame down to the following, keep all rows

#loan_amnt
#term
#int_rate
#emp_title
#annual_inc
#purpose

filtered_df = loans_csv.iloc[:, [0, 3, 4, 8, 11, 16]]
filtered_df.head()

loan_amnt       term  int_rate       emp_title  annual_inc  \
0      10000  36 months    0.1033             NaN    280000.0   
1       4000  36 months    0.2340        Security     90000.0   
2       5000  36 months    0.1797  Administrative     59280.0   
3       9600  36 months    0.1298             NaN     35704.0   
4       2500  36 months    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

### Filter Rows by 36 Month Terms

In [6]:
# Conditional indexing to filter DataFrame where 'term' is equal to '36 months'
term_df = filtered_df.loc[filtered_df["term"] == "36 months"].copy()
term_df.head()

loan_amnt       term  int_rate       emp_title  annual_inc  \
0      10000  36 months    0.1033             NaN    280000.0   
1       4000  36 months    0.2340        Security     90000.0   
2       5000  36 months    0.1797  Administrative     59280.0   
3       9600  36 months    0.1298             NaN     35704.0   
4       2500  36 months    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

### Modify Row Values

In [7]:
# Change row values within the 'term' column from '36 months' to '3 Years'
term_df.loc[term_df["term"] == "36 months", "term"] = "3 Years"
term_df.head()

loan_amnt     term  int_rate       emp_title  annual_inc  \
0      10000  3 Years    0.1033             NaN    280000.0   
1       4000  3 Years    0.2340        Security     90000.0   
2       5000  3 Years    0.1797  Administrative     59280.0   
3       9600  3 Years    0.1298             NaN     35704.0   
4       2500  3 Years    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

In [8]:
# Change row values within the 'emp_title' column from NaN to 'Unknown'
term_df.loc[term_df["emp_title"].isnull(), "emp_title"] = "Unknown"
term_df.head()

loan_amnt     term  int_rate       emp_title  annual_inc  \
0      10000  3 Years    0.1033         Unknown    280000.0   
1       4000  3 Years    0.2340        Security     90000.0   
2       5000  3 Years    0.1797  Administrative     59280.0   
3       9600  3 Years    0.1298         Unknown     35704.0   
4       2500  3 Years    0.1356            Chef     55000.0   

              purpose  
0  debt_consolidation  
1  debt_consolidation  
2  debt_consolidation  
3    home_improvement  
4  debt_consolidation

### View Summary Statistics for 3 Year Term Loans

In [9]:
# Describe summary statistics for three-year loans
term_df.describe(include="all")

loan_amnt     term      int_rate emp_title    annual_inc  \
count   88179.000000    88179  88179.000000     88179  8.817900e+04   
unique           NaN        1           NaN     31470           NaN   
top              NaN  3 Years           NaN   Unknown           NaN   
freq             NaN    88179           NaN     15669           NaN   
mean    13279.580456      NaN      0.120610       NaN  8.043616e+04   
std      9479.610905      NaN      0.048955       NaN  1.109251e+05   
min      1000.000000      NaN      0.060000       NaN  0.000000e+00   
25%      6000.000000      NaN      0.078400       NaN  4.500000e+04   
50%     10000.000000      NaN      0.110600       NaN  6.500000e+04   
75%     18000.000000      NaN      0.150200       NaN  9.500000e+04   
max     40000.000000      NaN      0.309900       NaN  9.000000e+06   

                   purpose  
count                88179  
unique                  12  
top     debt_consolidation  
freq                 45456  
mean                   NaN  
std                    NaN  
min                    NaN  
25%                    NaN  
50%                    NaN  
75%                    NaN  
max                    NaN

### Show Unique Values for Employee Titles with 3 Year Loans

In [10]:
# Calculate unique values and counts for employee titles of 3 year customer loans
term_df["emp_title"].value_counts()

Unknown                     15669
Teacher                      1433
Manager                      1207
Registered Nurse              672
Driver                        654
                            ...  
Distrbution Operator            1
Accountat                       1
Senior Strategic Analyst        1
Supply Chain Specialist         1
Ordnance Officer                1
Name: emp_title, Length: 31470, dtype: int64

### Show Unique Values for Purpose of 3 Year Loans

In [11]:
# Calculate unique values and counts for loan purposes of 3 year customer loans
term_df["purpose"].value_counts()

debt_consolidation    45456
credit_card           24780
other                  5836
home_improvement       5446
major_purchase         1795
medical                1245
car                     838
small_business          828
vacation                749
house                   596
moving                  547
renewable_energy         63
Name: purpose, dtype: int64

### Show Summary Statistics for 3 Year Loans of Customers with Annual Income Greater Than $80,000

In [12]:
# Display summary statistics where annual income is greater than $80,000 to find count and mean
term_df.loc[term_df["annual_inc"] > 80000].describe()

loan_amnt      int_rate    annual_inc
count  29748.000000  29748.000000  2.974800e+04
mean   18240.116142      0.114116  1.389634e+05
std    10991.909117      0.048067  1.752361e+05
min     1000.000000      0.060000  8.000100e+04
25%    10000.000000      0.075600  9.500000e+04
50%    15700.000000      0.104700  1.120000e+05
75%    25000.000000      0.144700  1.500000e+05
max    40000.000000      0.309400  9.000000e+06

### Show Summary Statistics for 3 Year Loans of Customers with Annual Income Less Than $80,000

In [13]:
# Display summary statistics where annual income is less than $80,000 to find count and mean
term_df.loc[term_df["annual_inc"] < 80000].describe()

loan_amnt      int_rate    annual_inc
count  56015.000000  56015.000000  56015.000000
mean   10607.487727      0.124268  49372.805964
std     7318.791900      0.049137  16669.098821
min     1000.000000      0.060000      0.000000
25%     5000.000000      0.081900  37000.000000
50%     9600.000000      0.113100  50000.000000
75%    14000.000000      0.150200  62635.500000
max    40000.000000      0.309900  79999.000000

### What insights can we answer regarding 3 year loans?

1. What kind of customers (employee title) seem to ask for three-year loans most frequently?
2. What are three-year loans generally used for?
3. What is the difference in count of three-year loan customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000?
4. What is the difference in interest rates of three-year loan customers with annual incomes greater than 80,000 compared to those with annual incomes less than 80,000? 

**Sample Answers**

1. Interestingly enough, Teachers and Managers seem to be the top two contenders for requesting three-year loans. Also, there is a disproportionate amount of 'Unknown' or un-reported employee titles. Unless this field is optional, this could signal a potential data quality problem.

2. From the value counts of loan purposes, it looks like three-year loans are generally used to consolidate debt.

3. A majority of three-year loan customers have annual incomes of less than 80,000. Counts of customers with annual incomes less than 80,000 compared to greater than 80,000 stands at 56,015 customers vs 29,748 customers, respectively.

4. Interest rates between customers with annual incomes higher than 
80,000 compared to those with less than 80,000 differ by approximately 1%; Interest rates were 12.43% and 12.41%, respectively.